router-tests/testenv/testenv.go

```diff
@@ -1910,68 +1914,176 @@ func (e *Environment) WaitForTriggerCount(desiredCount uint64, timeout time.Dura
+
+	for attempts < maxAttempts {
+		attempts++
+
+		// Set a reasonable write deadline
+		writeDeadline := time.Now().Add(2 * time.Second)
+		if err = conn.SetWriteDeadline(writeDeadline); err != nil {
+			t.Logf("Failed to set write deadline: %v", err)
+			return err
+		}
+
 		err = conn.WriteMessage(messageType, data)
-		if err != nil && strings.Contains(err.Error(), "connection reset by peer") {
-			t.Log("connection reset by peer found, retrying...")
-			err = conn.SetReadDeadline(time.Now().Add(1 * time.Second))
-			require.NoError(t, err)
-			time.Sleep(time.Duration(i*200) * time.Millisecond)
-			continue
+
+		// Reset the deadline to prevent future operations from timing out
+		if resetErr := conn.SetWriteDeadline(time.Time{}); resetErr != nil {
+			t.Logf("Failed to reset write deadline: %v", resetErr)
+		}
+
+		if err == nil {
+			return nil
 		}
-		break
+
+		if attempts >= maxAttempts {
+			return fmt.Errorf("failed to write message to WebSocket after %d attempts: %w", attempts, err)
+		}
+
+		// Get backoff duration for this attempt
+		duration := b.Duration()
+		t.Logf("WebSocket write message error (attempt %d/%d): %v - retrying in %v...",
+			attempts, maxAttempts, err, duration)
+
+		time.Sleep(duration)
 	}
 
-	return err
+	// This should not be reached
+	return fmt.Errorf("failed to write message to WebSocket: %w", err)
 }
 
-func DeflakeWSWriteJSON(t testing.TB, conn *websocket.Conn, v interface{}) (err error) {
-	for i := 0; i < 5; i++ {
+func WSWriteJSON(t testing.TB, conn *websocket.Conn, v interface{}) (err error) {
+	b := backoff.New(5*time.Second, 100*time.Millisecond)
+
+	attempts := 0
+	maxAttempts := 5
+
+	for attempts < maxAttempts {
+		attempts++
+
+		// Set a reasonable write deadline
+		writeDeadline := time.Now().Add(2 * time.Second)
+		if err = conn.SetWriteDeadline(writeDeadline); err != nil {
+			t.Logf("Failed to set write deadline: %v", err)
+			return err
+		}
+
 		err = conn.WriteJSON(v)
-		if err != nil && strings.Contains(err.Error(), "connection reset by peer") {
-			t.Log("connection reset by peer found, retrying...")
-			err = conn.SetReadDeadline(time.Now().Add(1 * time.Second))
-			require.NoError(t, err)
-			time.Sleep(time.Duration(i*200) * time.Millisecond)
-			continue
+
+		// Reset the deadline to prevent future operations from timing out
+		if resetErr := conn.SetWriteDeadline(time.Time{}); resetErr != nil {
+			t.Logf("Failed to reset write deadline: %v", resetErr)
+		}
+
+		if err == nil {
+			return nil
 		}
-		break
+
+		if attempts >= maxAttempts {
+			return fmt.Errorf("failed to write JSON to WebSocket after %d attempts: %w", attempts, err)
+		}
+
+		// Get backoff duration for this attempt
+		duration := b.Duration()
+		t.Logf("WebSocket write JSON error (attempt %d/%d): %v - retrying in %v...",
+			attempts, maxAttempts, err, duration)
+
+		time.Sleep(duration)
 	}
 
-	return err
+	// This should not be reached
+	return fmt.Errorf("failed to write JSON to WebSocket: %w", err)
```

Comment:
Just a nit, given you're already changing every call-site with the rename, if these functions get `t testing.TB`, and the `err` return is always just handled by `require.NoError`, why not just `t.Fail/t.Failnow` instead of returning errors? Did I miss a `require.Error/ErrorContains` somewhere?

Reply:
Because it is a low level function. I don't want to assume something. Maybe people want to check for the error.